src/Workspaces/Core/Portable/Rename/Renamer.RenameSymbolDocumentAction.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Rename
+{
+    public static partial class Renamer
+    {
+        /// <summary>
+        /// Action that will rename a type to match the current document name
+        /// </summary>
+        internal sealed class RenameSymbolDocumentAction : RenameDocumentAction
+        {
+            private readonly AnalysisResult _analysis;
+
+            private RenameSymbolDocumentAction(
+                AnalysisResult analysis,
+                OptionSet optionSet,
+                ImmutableArray<ErrorResource> errors)
+                : base(errors, optionSet)
+            {
+                _analysis = analysis;
+            }
+
+            public override string GetDescription(CultureInfo? culture)
+                => string.Format(WorkspacesResources.ResourceManager.GetString("Rename_0_to_1", culture ?? WorkspacesResources.Culture)!, _analysis.OriginalDocumentName, _analysis.NewDocumentName);
+
+            internal override async Task<Solution> GetModifiedSolutionAsync(Document document, CancellationToken cancellationToken)
+            {
+                var solution = document.Project.Solution;
+                var matchingTypeDeclaration = await GetMatchingTypeDeclarationAsync(document, _analysis.OriginalSymbolName!, cancellationToken).ConfigureAwait(false);
+
+                if (matchingTypeDeclaration is object)
+                {
+                    var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                    var symbol = semanticModel.GetDeclaredSymbol(matchingTypeDeclaration, cancellationToken);
+
+                    solution = await RenameSymbolAsync(solution, symbol, _analysis.NewSymbolName, OptionSet, cancellationToken).ConfigureAwait(false);
+                }
+
+                return solution;
+            }
+
+            private static async Task<SyntaxNode?> GetMatchingTypeDeclarationAsync(Document document, string name, CancellationToken cancellationToken)
+            {
+                var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+                var typeDeclarations = syntaxRoot.DescendantNodesAndSelf().Where(syntaxFacts.IsTypeDeclaration);
```

Comment:
Can we tweak the DescendantNodesAndSelf so it doesn't walk into method bodies? Something like:

```suggestion
                var typeDeclarations = syntaxRoot.DescendantNodesAndSelf(n => !syntaxFacts.IsMethodBody(n)).Where(syntaxFacts.IsTypeDeclaration);
```

might be sufficient?

Reply:
good call